framework/src/modules/bft/bft_votes.ts

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright © 2021 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ */
+
+import { BlockHeader } from '../../node/state_machine';
+import { BFTParametersCache } from './bft_params';
+import { BFTVotes } from './schemas';
+import { getBlockBFTProperties } from './utils';
+
+export const insertBlockBFTInfo = (
+	bftVotes: BFTVotes,
+	header: BlockHeader,
+	maxLength: number,
+): void => {
+	// eslint-disable-next-line no-param-reassign
+	bftVotes.blockBFTInfos = [getBlockBFTProperties(header), ...bftVotes.blockBFTInfos].slice(
+		0,
+		maxLength,
+	);
+};
+
+export const getHeightNotPrevoted = (bftVotes: BFTVotes): number => {
+	const [newBlockBFTInfo] = bftVotes.blockBFTInfos;
+	const { height: currentHeight } = newBlockBFTInfo;
+	let heightPreviousBlock = newBlockBFTInfo.maxHeightGenerated;
+
+	// iterate over blockBFTInfo objects in decreasing order by height
+	while (currentHeight - heightPreviousBlock < bftVotes.blockBFTInfos.length) {
+		const blockBFTInfo = bftVotes.blockBFTInfos[currentHeight - heightPreviousBlock];
+		if (
+			!blockBFTInfo.generatorAddress.equals(newBlockBFTInfo.generatorAddress) ||
+			blockBFTInfo.maxHeightGenerated >= heightPreviousBlock
+		) {
+			return heightPreviousBlock;
+		}
+		heightPreviousBlock = blockBFTInfo.maxHeightGenerated;
+	}
+	// return one less than the height of the blockBFTInfo with smallest height
+	return bftVotes.blockBFTInfos[bftVotes.blockBFTInfos.length - 1].height - 1;
+};
+
+export const updatePrevotesPrecommits = async (
+	bftVotes: BFTVotes,
+	paramsCache: BFTParametersCache,
+): Promise<void> => {
+	// when processing the genesis block the array is empty and no prevotes or precommits need to be updated
+	if (bftVotes.blockBFTInfos.length === 0) {
+		return;
+	}
+	const [newBlockBFTInfo] = bftVotes.blockBFTInfos;
+	// a block header only implies votes if maxHeightGenerated < height
+	if (newBlockBFTInfo.maxHeightGenerated >= newBlockBFTInfo.height) {
+		return;
+	}
+	const validatorInfo = bftVotes.activeValidatorsVoteInfo.find(v =>
+		v.address.equals(newBlockBFTInfo.generatorAddress),
+	);
+	if (!validatorInfo) {
```

Comment:
If the validator is not contained in the active validators, we can just return here. This will happen, for example, for standby delegates which have BFT weight 0 and are therefore not contained in the active validators (actually, the API `setBFTParameters` only allows positive BFT weights). 

I missed this point in the BFT module LIP and will also update it there.